src/Castle.Core/DynamicProxy/CustomAttributeInfo.cs

```diff
@@ -0,0 +1,294 @@
+			FieldInfo[] namedFields,
+			object[] fieldValues)
+			: this(con, constructorArgs, EmptyProperties, EmptyValues, namedFields, fieldValues)
+		{
+		}
+
+		public CustomAttributeInfo(
+			ConstructorInfo con,
+			object[] constructorArgs)
+			: this(con, constructorArgs, EmptyProperties, EmptyValues, EmptyFields, EmptyValues)
+		{
+		}
+
+		public static CustomAttributeInfo FromExpression(Expression<Func<Attribute>> expression)
+		{
+			var namedProperties = new List<PropertyInfo>();
+			var propertyValues = new List<object>();
+			var namedFields = new List<FieldInfo>();
+			var fieldValues = new List<object>();
+
+			var newExpression = expression.Body as NewExpression;
+			if (newExpression == null)
+			{
+				var memberInitExpression = expression.Body as MemberInitExpression;
+				if (memberInitExpression == null)
+				{
+					throw new ArgumentException("The expression must be either a simple constructor call or an object initializer expression");
+				}
+
+				newExpression = memberInitExpression.NewExpression;
+
+				foreach (var binding in memberInitExpression.Bindings)
+				{
+					var assignment = binding as MemberAssignment;
+					if (assignment == null)
+					{
+						throw new ArgumentException("Only assignments bindings are supported");
+					}
+
+					object value = GetAttributeArgumentValue(assignment.Expression, true);
+
+					var property = assignment.Member as PropertyInfo;
+					if (property != null)
+					{
+						namedProperties.Add(property);
+						propertyValues.Add(value);
+					}
+					else
+					{
+						var field = assignment.Member as FieldInfo;
+						if (field != null)
+						{
+							namedFields.Add(field);
+							fieldValues.Add(value);
+						}
+						else
+						{
+							throw new ArgumentException("Only property and field assignments are supported");
+						}
+					}
+				}
+			}
+
+			var ctorArguments = new List<object>();
+			foreach (var arg in newExpression.Arguments)
+			{
+				object value = GetAttributeArgumentValue(arg, true);
+				ctorArguments.Add(value);
+			}
+
+			return new CustomAttributeInfo(
+				newExpression.Constructor,
+				ctorArguments.ToArray(),
+				namedProperties.ToArray(),
+				propertyValues.ToArray(),
+				namedFields.ToArray(),
+				fieldValues.ToArray());
+		}
+
+		private static object GetAttributeArgumentValue(Expression arg, bool allowArray)
+		{
+			var constant = arg as ConstantExpression;
+			if (constant == null)
+			{
+				if (allowArray)
+				{
+					var newArrayExpr = arg as NewArrayExpression;
+					if (newArrayExpr != null)
+					{
+						var array = new object[newArrayExpr.Expressions.Count];
+						int index = 0;
+						foreach (var expr in newArrayExpr.Expressions)
+						{
+							object value = GetAttributeArgumentValue(expr, false);
+							array[index] = value;
+							index++;
+						}
+						return array;
+					}
+				}
+				throw new ArgumentException("Only constant and single-dimensional array expressions are supported");
```

Comment:
We may reach this line in a case where a single-dimensional array expression is _not_ supported. (When `allowArray` is `false`.)
I'd consider reworking the flow to avoid a potentially misleading exception message.
While I'm commenting on it, if such a rework avoided a boolean parameter, I'd consider it an improvement. I never know what these mean. Way back on line 113, I was all '"true"? What's "true" mean?".
But it's not such a huge deal, given that it's a private method and we can find the answer pretty quickly.